### PR TITLE
C#: Adopt shared SSA data-flow integration

### PR DESCRIPTION
**Pull request by hvitved**
This PR adopts the newly introduced shared [SSA data-flow integration layer](https://github.com/github/codeql/pull/16884). A side-effect is that we get [phi-input barrier guards](https://github.com/github/codeql/pull/15985) for free, which had not previously been ported to C#.

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPublic.qll [hunk 1]
```diff
@@ -171,8 +171,14 @@ signature predicate guardChecksSig(Guard g, Expr e, AbstractValue v);
...
   /** Gets a node that is safely guarded by the given guard check. */
-  ExprNode getABarrierNode() {
+  pragma[nomagic]
+  Node getABarrierNode() {
+    SsaFlow::asNode(result) =
+      SsaImpl::DataFlowIntegration::BarrierGuard<guardChecks/3>::getABarrierNode()
+    or
     exists(Guard g, Expr e, AbstractValue v |
       guardChecks(g, e, v) and
       g.controlsNode(result.getControlFlowNode(), e, v)
```
Comment (aschackmull) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPublic.qll:184]: Why is this second disjunct not redundant now? Does it cover more than ssa variable reads?

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPublic.qll:184]: Good question: Yes, it covers more than SSA variable reads (it uses GVN, which--although it is unsound in general--is likely to produce valid guards).

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -827,23 +750,14 @@ predicate simpleLocalFlowStep(Node nodeFrom, Node nodeTo, string model) {
...
+      isUseStep = false
       or
-      // Flow into phi (read)/uncertain SSA definition node from read
-      exists(Node read | LocalFlow::localFlowSsaInputFromRead(read, def, nodeTo) |
-        nodeFrom = read and
-        not FlowSummaryImpl::Private::Steps::prohibitsUseUseFlow(nodeFrom, _)
-        or
-        nodeFrom.(PostUpdateNode).getPreUpdateNode() = read
-      )
+      not FlowSummaryImpl::Private::Steps::prohibitsUseUseFlow(nodeFrom, _)
```
Comment (aschackmull) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:760]: There's a pipeline duplication here, better add `isUseStep = true` to the second disjunct.

**Review by aschackmull — APPROVED** [review]
Looks plausible to me.

**Pull request merged**